### PR TITLE
fix(preprocessing): preserve segmentation mask dtype across augmentation layers

### LINKED ISSUE #20857: Several Problems when using Keras Image Augmentation Layers to Augment Images and Masks for a Semantic Segmentation Dataset
In the following, I will try to tell you a bit of my “story” about how I read through the various layers and (partly outdated) tutorials on image augmentation for semantic segmentation, what problems and bugs I encountered and how I was able to solve them, at least for my use case. I hope that I am reporting these issue(s) in the right place and that I can help the developers responsible with my “story”.

I am currently using the latest versions of Keras (3.8) and Tensorflow (2.18.0) on an Ubuntu 23.10 machine, but have also observed the issues with other setups.

My goal was to use some of the Image Augmentation Layers from Keras to augment the images AND masks in my semantic segmentation dataset. To do this I followed the [segmentation tutorial from Tensarflow](https://www.tensorflow.org/tutorials/images/segmentation) where a custom `Augment` class is used for augmentation.

This code works fine because the `Augment` class uses only a single augmentation layer. But, for my custom code I wanted to use more augmentations. Therefore, I "combined" several of the augmentation layers trying both the normal [Pipeline layer](https://keras.io/api/layers/preprocessing_layers/image_augmentation/pipeline/) but also my preferred [RandomAugmentationPipeline layer from keras_cv](https://github.com/keras-team/keras-cv/blob/master/keras_cv/src/layers/preprocessing/random_augmentation_pipeline.py).

I provide an example implementation in [this gist](https://colab.research.google.com/gist/RabJon/9f27d600301dcf6a734e8b663c802c7f/modification_of_segmentation.ipynb). The gist is a modification of the aforementioned [image segmentation tutorial from Tensorflow](https://www.tensorflow.org/tutorials/images/segmentation) where I intended to only change the `Augment` class slightly to fit my use case and to remove all the unnecessary parts. Unfortunately, I also had to touch the code that loads the dataset, because the dataset version used in the notebook was no longer supported. But, this is another issue. 

Anyway, at the bottom of the notebook you can see some example images and masks from the data pipeline and if you inspect the in more detail you can see that for some examples images and masks do not match because they were augmented differently, which is obviously not the expected behaviour. You can also see that these mismatches already happen from the first epoch on. On my system with my custom use case (where I was using the RandomAugmentationPipeline layer) it happend only from the second epoch onward, which made debugging much more difficult.

I first assumed that it was one specific augmentation layer that caused the problems, but after trying them one-by-one I found out that it is the combination of layers that makes the problems. So, I started to think about possible solutions and I also tried to use the Sequential model from keras to combine the layers instead of the Pipeline layer, but the result remained the same.

I found that the random seed is the only parameter which can be used to "control and sync" the augmentation of the images and masks, but I already had ensured that I always used the same seed. So, I started digging into the source code of the different augmentation layers and I found that most of the ones that I was using implement a `transform_segmentation_masks()` method which is however not used if the layers are called as described in the tutorial. This is because, in order to enforce the call of this method, images and masks must be passed to the same augmentation layer object as a dictionary with the keys “images” and “segmentation_masks” instead of using two different augmentation layer objects for images and masks. However, I had not seen this type of call using a dictionary in any tutorial, neither for Tensorflow nor for Keras. Nevertheless, I decided to change my code to make use of the `transform_segmentation_masks()` method, as I hoped that if such a method already existed, it would also process the images and masks correctly and thus avoid the mismatches. 
Unfortunately, this was not directly the case, because some of the augmentation layers changed the masks to float32 data types although the input was uint8. Even with layers such as “RandomFlip”, which should not change or interpolate the data at all, but only shift it. So, I had to wrap all layers again in a custom layer, which castes the data back to the input data type before returning it:

```python
class ApplyAndCastAugmenter(BaseImageAugmentationLayer):
    
    def __init__(self, other_augmenter, **kwargs):
        super().__init__(**kwargs)
        
        self.other_augmenter = other_augmenter
    
    def call(self, inputs):
        
        output_dtypes = {}
        is_dict = isinstance(inputs, dict)
        if is_dict:
            for key, value in inputs.items():
                output_dtypes[key] = value.dtype
        else:
            output_dtypes = inputs.dtype
        
        outputs = self.other_augmenter(inputs)
        
        if is_dict:
            for key in outputs.keys():
                outputs[key] = keras.ops.cast(outputs[key], output_dtypes[key])
        else:
            outputs = keras.ops.cast(outputs, output_dtypes)
        
        return outputs
```

With this (admittedly unpleasant) work-around, I was able to fix all the remaining problems. Images and masks finally matched perfectly after augmentation and this was immediately noticeable in a significant performance improvement in the training of my segmentation model.

For the future I would now like to see that wrapping with my custom `ApplyAndCastAugmenter` is no longer necessary but is handled directly by the `transform_segmentation_masks()` method and it would also be good to have a proper tutorial on image augmentation for semantic segmentation or to update the old tutorials.

### PR DESCRIPTION
**Pull request by rstar327**
## Description

Preserves the dtype of `segmentation_masks` when running them through any image preprocessing layer. Fixes #20857.

The bug: when a layer dict-call includes `segmentation_masks`, the per-layer `transform_segmentation_masks` implementations typically delegate to `transform_images`, which casts inputs to the layer's `compute_dtype` (usually `float32`). As a result, a `uint8` class-index mask comes back as a `float32` tensor — silently breaking any downstream code that expects integer-class indices.

### Reproducer (master, before fix)

```python
mask = np.array([[[[0],[1],[0]],[[1],[1],[0]]]], dtype="uint8")
layer = keras.layers.RandomFlip("horizontal", seed=0)
out = layer({"images": mask, "segmentation_masks": mask})
print(out["segmentation_masks"].dtype)  # -> float32  (BUG)
```

### Fix

`BaseImagePreprocessingLayer.call` now captures the input mask dtype before delegating to the per-layer transform and casts the result back if the dtype changed. The fix is generic (covers every subclass of `BaseImagePreprocessingLayer`) and adds a single comparison + conditional cast on the mask path — no impact on the image / label / bounding-box paths.

### Tests

Added `test_segmentation_mask_dtype_is_preserved` to `base_image_preprocessing_layer_test.py`. Passes on tensorflow, jax, torch, numpy backends locally. Full existing preprocessing test suite still green (44 tests in the surrounding modules).

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

**Review by hertschuh — CHANGES_REQUESTED** [review]
First, I thought this issue had already been addressed. Which preprocessing layers still have the issue?

Second, we should use a different approach. The faulty layer classes should be addressed individually.

Here is why. Let's say you have a image with mask `2`, and mask `4` used. Let's say you do a rotation or rescaling and because of the interpolation, we now have float values between `2` and `4`. After casting, we'll have some mask values of `3`. But mask `3` didn't exist in the first place. We should not let new mask values appear from nowhere.

This will force us to verify none of the layers use interpolation for masks. I thought https://github.com/keras-team/keras/pull/22192 covered this, but maybe some were missed.

**Comment by rstar327** [comment]
Closing in favor of #23191, which uses the per-layer approach you asked for, @hertschuh.

Still-faulty layers fixed there: `CenterCrop`/`RandomCrop` (resize interpolates masks → invented classes, plus `float32` cast) and `RandomFlip`/`CutMix` (exact ops but `float32` cast). `MixUp`/`AugMix` blend, which is undefined for class-index masks — left as a separate decision.
